src/MultiSigSigner.sol

```diff
@@ -0,0 +1,179 @@
+
+        Config storage config = _configs[msg.sender][keyHash];
+
+        if (config.threshold > 0) {
+            revert ConfigAlreadySet();
+        }
+
+        _configs[msg.sender][keyHash] =
+            Config({threshold: threshold, ownerKeyHashes: ownerKeyHashes});
+    }
+
+    function addOwner(bytes32 keyHash, bytes32 ownerKeyHash) public {
+        _checkKeyHash(keyHash);
+
+        Config storage config = _configs[msg.sender][keyHash];
+        config.ownerKeyHashes.push(ownerKeyHash);
+    }
+
+    function removeOwner(bytes32 keyHash, bytes32 ownerKeyHash) public {
+        _checkKeyHash(keyHash);
+
+        Config storage config = _configs[msg.sender][keyHash];
+        bytes32[] storage ownerKeyHashes_ = config.ownerKeyHashes;
+        uint256 ownerKeyCount = ownerKeyHashes_.length;
+
+        bool found = false;
+
+        for (uint256 i = 0; i < ownerKeyCount; i++) {
+            if (ownerKeyHashes_[i] == ownerKeyHash) {
+                // Replace the owner to remove with the last owner
+                ownerKeyHashes_[i] = ownerKeyHashes_[ownerKeyCount - 1];
+                // Remove the last element
+                ownerKeyHashes_.pop();
+                found = true;
+                break;
+            }
+        }
+
+        if (!found) revert OwnerNotFound();
+        if (ownerKeyCount - 1 < config.threshold) revert InvalidThreshold();
+    }
+
+    function setThreshold(bytes32 keyHash, uint256 threshold) public {
+        _checkKeyHash(keyHash);
+
+        Config storage config = _configs[msg.sender][keyHash];
+
+        if (threshold == 0 || threshold > config.ownerKeyHashes.length) revert InvalidThreshold();
+
+        config.threshold = threshold;
+    }
+
+    ////////////////////////////////////////////////////////////////////////
+    // Signature Validation
+    ////////////////////////////////////////////////////////////////////////
+
+    /// @dev This function SHOULD only be called by valid Delegation porto accounts.
+    /// - This will iteratively make a call to the address(msg.sender).unwrapAndValidateSignature
+    ///   for each owner key hash in the config.
+    /// - Signature of a multi-sig should be encoded as abi.encode(bytes[] memory ownerSignatures)
+    /// - For efficiency, place the signatures in the same order as the ownerKeyHashes in the config.
+    /// - Failing owner signatures are ignored, as long as valid signaturs > threshold.
+    function isValidSignatureWithKeyHash(bytes32 digest, bytes32 keyHash, bytes memory signature)
+        public
+        view
+        returns (bytes4 magicValue)
+    {
+        bytes[] memory signatures = abi.decode(signature, (bytes[]));
+        Config memory config = _configs[msg.sender][keyHash];
+
+        uint256 validKeyNum;
+
+        for (uint256 i; i < signatures.length; ++i) {
+            (bool isValid, bytes32 ownerKeyHash) =
+                IDelegation(msg.sender).unwrapAndValidateSignature(digest, signatures[i]);
```

Comment:
Note to self:

This MultiSigSigner requires the `msg.sender` to implement Delegation.